osg-3.6 with Mak patches applied

john-terrell opened the pull request:
This Pull Request contains Mak's patches to osg 3.6.

john-terrell committed:
Merge of Mak's OSG 3.4 changes.   * Untested *

john-terrell committed:
Updated to latest SVN drop.

john-terrell commented:
NOTE: The PR for this branch is being updated to fix build breaks caused by the OSG 3.4 -> 3.6 merge.   For changes that need review by others, this is where those changes will be reviewed.

john-terrell committed:
Update FindNVTT.cmake

john-terrell committed:
Update GLExtensions

john-terrell committed:
Update Texture

john-terrell committed:
Update Program.cpp

john-terrell committed:
Update Shader.cpp

john-terrell committed:
Update Shader.cpp

john-terrell committed:
Update FrameBufferObject.cpp

john-terrell committed:
Build Fixes.

john-terrell committed:
Build break updates.

john-terrell committed:
More merge fixes.

john-terrell committed:
More merge fixes

john-terrell committed:
Merge fixes.

john-terrell committed:
Merge fixes

john-terrell committed:
Merge fixes.

john-terrell commented:
OPEN ISSUE: osgParticle has been reworked for OSG 3.6.   Need to revisit with Gedalia.

john-terrell commented:
OPEN ISSUE: ShaderGenCache no longer exists.   Need to understand if it was used in Vantage and if so, is there any changes that need to be reflected to account for it.

john-terrell committed:
Merge fixes.

john-terrell committed:
Merge fixes

john-terrell committed:
Merge fixes.

john-terrell commented:
OPEN ISSUE: fbx plugin for OSG3.6 was COMPLETELY overwritten by Mak's version due to the updated FBX SDK (2013 vs 2019) it uses.

john-terrell committed:
Merge fixes

john-terrell commented:
> OPEN ISSUE: ShaderGenCache no longer exists. Need to understand if it was used in Vantage and if so, is there any changes that need to be reflected to account for it.

I've replaced all of Mak's old OSG3.4 ShaderGen source with OSG 3.6's ShaderGen.    Will verify that no code in Mak relies on the old Mak ShaderGen.

john-terrell commented:
src\osgPlugins\OpenFlight's OSG 3.6 changes replaced by Mak's OSG3.4.

john-terrell committed:
Merge fixes

john-terrell committed:
Merge fixes.

john-terrell committed:
plavoie - Feature #5189 and some scaling fix for feature #5092

john-terrell committed:
plavoie - Trying to fix FBX and Boost using for VC12

john-terrell committed:
plavoie - (FBX fix) Move a function to the fbxUtil Class .. and remove the usage of a boost filesystem V3 function

john-terrell committed:
plavoie - FBX fix for CSV texture list not persistent after first load

john-terrell committed:
thogarth - added ovr_multiview support to FrameBufferObject and const to camera to enable the extension on attachments

john-terrell committed:
tgeorge - Defect 66340 - Changed the object mutex to mutable so const objects can still access it.
https://ontime/viewitem.aspx?id=66340&type=defects&force_use_number=false

john-terrell committed:
plavoie - Fix some pivot points for some wheels (not the one from Maya)
Fix the Defect#66674 - some FBX models output the following warning (Warning: TexEnvCombine::apply(State&) - not supported.)

john-terrell committed:
plavoie - Extended the feature #5092 to support custom node name with the CSV file

john-terrell committed:
tgeorge - Defect 66708 - Move make current up so it is called before the cull thread starts. This prevents the main thread optional jobs from being starved in the case where make current takes longer than cull. This should not impact the single window case.
https://ontime/viewitem.aspx?id=66708&type=defects&force_use_number=false

john-terrell committed:
pprabhu - -added getCurrentContext

john-terrell committed:
Updated GL3_FEATURES check to accomodate the older compilers.

john-terrell committed:
Specify MSVC_VER > 1600 for explicit operator on SafeArray.

john-terrell committed:
Merge branch 'mak/osg-3.4-merge' of https://github.com/VTMAK/OpenSceneGraph into mak/osg-3.4-merge

john-terrell committed:
Undo change yesterday for VC10 (which broken Linux).   VC10 won't be built anymore.

john-terrell committed:
Updated Material.cpp to fix merge issue that caused material spew.

john-terrell committed:
Fixed issue with OSG variable substitution (merge error)

john-terrell committed:
Fixed issue with OSG3.6 not checking for dirty texture parameters before attempting to apply them.   This was causing GL errors to be emitted related to immutable textures.

tcarellasMAK committed:
Fixed install issues and paths

john-terrell committed:
skoren - Commit makOsg ap-ubolight-stripe-staging to trunk.

john-terrell committed:
skoren - Change snprintf to sprintf: snprintf is not available in all build environments

john-terrell committed:
pprabhu - fixes for 66933: Encapsulate glPushDebugGroup/glPopDebug group in scoped variable, and be able to turn it off

john-terrell committed:
pprabhu - minor

john-terrell committed:
thogarth - merged multiview branch to add multisampling support to texture2d arrays

john-terrell committed:
Removed call that was missed by commit 710b09d0946daa0943e395fef9bc06c533583602.

john-terrell committed:
thogarth - fix for linux compile issue

john-terrell committed:
thogarth - fix for wrong case in include path

john-terrell committed:
Removed unsupported OSG 3.4 TextureObjectMangaer calls.

john-terrell committed:
gpasternak - starting to add tracy

john-terrell committed:
gpasternak - fixed scope

john-terrell committed:
gpasternak - linux fix maybe.

john-terrell committed:
gpasternak - another profile linux fix

john-terrell committed:
gpasternak - tracy markers

john-terrell committed:
gpasternak - markers, compile botch

john-terrell committed:
gpasternak - init file

john-terrell committed:
gpasternak - more profile include headers

john-terrell committed:
gpasternak - linux fix

john-terrell committed:
gpasternak - render stage marker

john-terrell committed:
Fixed broken merge insdie Renderer::cull_draw()

john-terrell committed:
tcarellas (rev 206573) - Added more logic for checking if TRACY is enabled

john-terrell committed:
gpasternak (Rev 206587) - bump tracy to have #define TRACY_DELAYED_INIT

john-terrell committed:
gpasternak (Rev 206626) - #def tweaks

john-terrell committed:
pzimmons (Rev 207680) - It looks like glUseProgram was being called multiple times even though VirtualProgram has a check to call glUseProgram once. It was due to a missed flag reset in State::getDefineString.

john-terrell committed:
pzimmons (Rev 207751) - An update to the last fix. A little more sophisticated.

john-terrell committed:
pzimmons (Rev 207759) - Reverting the glUseProgram fixes since it seems to mess up the vegetation on osgEarth. Maybe there is something missing in these changes. The performance gain is good so it should be pursued.

john-terrell committed:
gpasternak (Rev 207915) - add profile directory to osg in general, profile texture uploads

john-terrell committed:
pzimmons (Rev 208011) - Putting back the glUseProgram change (to call glUseProgram a lot less frequently - only when Virtual Program agrees it should update, or when a system like triton or indirect needs to do it) since it is not related to the vegetation going underground.

I see Makland 117-145 -> 130->160 and California go from the 85s ftps to bumping into 90fps

john-terrell committed:
pzimmons (Rev 208173) - Adding in some tracey profile markers to show when we are mip mapping in the driver during render (mostly due to osgEarth elevation R32F textures and normal maps). About 4ms each time.

john-terrell committed:
gpasternak (Rev 208367) - added wrapper around zone text

john-terrell committed:
tgeorge (Rev 208418) - Defect 67882 - Fixed the max number of graphics contexts to stay correct even after osgEarth lowers the count when a graphics context is destroyed.

john-terrell committed:
tgeorge (Rev 208652) - Defect 67935 - Added a mutex to the fbx loader to prevent loading two fbx models at the same time.

john-terrell committed:
Fixed OsgProfiletext() at wrong spot in header.

tgeorgeVTMak commented on include/osg/GLDefines:
```diff
@@ -566,6 +566,7 @@ typedef char GLchar;
 #define GL_UNSIGNED_INT64_VEC2_ARB 0x8FF5
 #define GL_UNSIGNED_INT64_VEC3_ARB 0x8FF6
 #define GL_UNSIGNED_INT64_VEC4_ARB 0x8FF7
+
```
Did include/osg/GLBeginEndAdapter get removed completely and is no longer used by the particle system?

tgeorgeVTMak commented on include/osg/KdTree:
```diff
@@ -212,7 +212,6 @@ class OSG_EXPORT KdTreeBuilder : public osg::NodeVisitor
     protected:
 
         virtual ~KdTreeBuilder() {}
-
```
Can we go through and remove any changes that are whitespace only? No point in dragging them forward and cluttering the diffs between the master osg 3.6 branch and our makOsg branch.

tgeorgeVTMak replied:
I wouldn't worry about the whitespace changes if there are other changes in the same file. Just when the only change in a file is whitespace only.

tgeorgeVTMak commented on include/osg/Program:
```diff
@@ -252,7 +252,19 @@ class OSG_EXPORT Program : public osg::StateAttribute
         typedef std::map<std::string, UniformBlockInfo> UniformBlockMap;
 
         //const UniformBlockMap& getUniformBlocks(unsigned contextID) const;
-    public:
```
This file is missing a change to use flat_map. Might have been added after the last time you merged in makOsg.

tgeorgeVTMak commented on include/osg/State:
```diff
@@ -42,6 +42,10 @@
 #define GL_TEXTURE0 0x84C0
 #endif
 
+//VRV_PATCH to help get occlusion culling to work
```
This file also has a flat_map for faster uniforms change that is missing.

tgeorgeVTMak commented on include/osg/Texture:
```diff
@@ -652,11 +675,9 @@ class OSG_EXPORT Texture : public osg::StateAttribute
         inline InternalFormatType getInternalFormatType() const { return _internalFormatType; }
 
-        /* select the size internal format to use based on Image when available or Texture format settings.*/
-        GLenum selectSizedInternalFormat(const osg::Image* image=0) const;
```
Curious why this was removed. Maybe the cpp changes will explain it but I am not seeing this removed from the 3.4 diffs.

tgeorgeVTMak commented on include/osgParticle/ParticleSystem:
```diff
@@ -265,7 +265,7 @@ namespace osgParticle
           * for all graphics contexts. */
         virtual void releaseGLObjects(osg::State* state=0) const;
 
-        virtual osg::VertexArrayState* createVertexArrayStateImplementation(osg::RenderInfo& renderInfo) const;
```
This file is missing some changes:
setNumParticlesToCreateHint
   // BEGIN VRV_PATCH
        // don't compute these each time, just store in class
        const osg::Vec3 _xAxis;
        const osg::Vec3 _yAxis;
        const osg::Vec3 _zAxis;
        const Particle::Shape _userShape;
        // END VRV_PATCH

tgeorgeVTMak commented on src/osg/CMakeLists.txt:
```diff
@@ -76,6 +84,7 @@ SET(TARGET_H
     ${HEADER_PATH}/DrawPixels
     ${HEADER_PATH}/Endian
     ${HEADER_PATH}/Export
+    ${HEADER_PATH}/ExtendedMaterial
```
File is missing flat_map.cpp addition.

tgeorgeVTMak commented on src/osg/ContextData.cpp:
```diff
@@ -163,7 +163,12 @@ unsigned int ContextData::createNewContextID()
 
             OSG_INFO<<"ContextData::createNewContextID() : reusing contextID="<<itr->first<<std::endl;
 
-            return itr->first;
```
(comment about missing file not this file) The file CollectOcludersVisitor.cpp was changed for osg 3.4 but that change is not here.

tgeorgeVTMak replied:
Nevermind, I see that it was fixed in osg proper.

tgeorgeVTMak commented on src/osg/ContextData.cpp:
```diff
@@ -163,7 +163,12 @@ unsigned int ContextData::createNewContextID()
+            // update the maximum number of graphics contexts,
+            // to ensure that texture objects and display buffers are configured to the correct size.
+            osg::DisplaySettings::instance()->setMaxNumberOfGraphicsContexts(s_contextIDMap.size()); 
+            //END VRV_PATCH
```
note to future reviewers. This was in GraphicsContext.cpp in 3.4 and it has moved to here and this patch is correct.

tgeorgeVTMak commented on src/osg/DisplaySettings.cpp:
```diff
@@ -240,7 +241,7 @@ void DisplaySettings::setDefaults()
     _implicitBufferAttachmentRenderMask = DEFAULT_IMPLICIT_BUFFER_ATTACHMENT;
     _implicitBufferAttachmentResolveMask = DEFAULT_IMPLICIT_BUFFER_ATTACHMENT;
-    _glContextVersion = OSG_GL_CONTEXT_VERSION;
+    _glContextVersion = "1.0";
```
Why was this changed back to "1.0"?

tgeorgeVTMak commented on src/osg/Drawable.cpp:
```diff
@@ -226,16 +227,15 @@ Drawable::Drawable()
-    _useVertexBufferObjects = true;
-#endif
+    _supportsVertexBufferObjects = true; // OSG 3.6
+    _useVertexBufferObjects = false;  // MAK OSG 3.4 (OSG 3.6 = true)
```
It looks like we set the default to be on only for Oculus but maybe we can turn it on for everything now and get rid of the s_useVertexBufferObjectsDefault and the setter function for that. I guess lets leave it off for the initial 3.6 merge but put in a defect to ontime for trying to turn it on all the time and see if it works.

tgeorgeVTMak commented on src/osg/FrameBufferObject.cpp:
```diff
@@ -21,6 +21,7 @@
 #include <osg/Texture2DMultisample>
 #include <osg/Texture3D>
 #include <osg/Texture2DArray>
+#include <osg/Texture2DMultisampleArray>
```
Missing change to flushDeletedRenderBuffers() the GLExtensions::Get() function is false in makOsg for 3.4

tgeorgeVTMak replied:
Also what happened to flushDeletedFrameBufferObjects?

tgeorgeVTMak commented on src/osg/GLExtensions.cpp:
```diff
@@ -449,6 +449,9 @@ GLExtensions::GLExtensions(unsigned int in_contextID):
     glVersion = validContext ? findAsciiToFloat( versionString ) : 0.0f;
     glslLanguageVersion = 0.0f;
 
+    // vantage change
```
Comment for missing file. What happened to GLBeginEndAdapter.cpp?

tgeorgeVTMak commented on src/osg/GraphicsContext.cpp:
```diff
@@ -239,7 +243,7 @@ GraphicsContext::Traits::Traits(DisplaySettings* ds):
             swapBarrier(0),
             useMultiThreadedOpenGLEngine(false),
             useCursor(true),
-            glContextVersion(OSG_GL_CONTEXT_VERSION),
```
Same as the other place, why are we not taking this change?

tgeorgeVTMak commented on src/osg/LightModel.cpp:
```diff
@@ -53,37 +53,41 @@ LightModel::~LightModel()
 void LightModel::apply(State&) const
 {
-    OSG_NOTICE<<"Warning: LightModel::apply(State&) - not supported."<<std::endl;
+    static int i = 1;
```
If it is easy to remove all the other changes that are whitespace only in this file we should do that. Really it is just the static int i = 1 changes here that are our changes.

tgeorgeVTMak commented on src/osg/Program.cpp:
```diff
@@ -659,6 +705,10 @@ Program::PerContextProgram::PerContextProgram(const Program* program, unsigned i
 
         if (_glProgramHandle)
         {
+            if (_extensions->glObjectLabel && program->getName().length()) {
```
put //VRV_PATCH here (it is there in the makOsg 3.4)

tgeorgeVTMak commented on src/osg/Shader.cpp:
```diff
@@ -407,36 +407,34 @@ Shader::ShaderObjects::ShaderObjects(const osg::Shader* shader, unsigned int con
 
```
Missing faster shader define stuff at the top. Probably newer change than the last time makOsg was synced with this branch.

tgeorgeVTMak commented on src/osg/Shader.cpp:
```diff
@@ -407,36 +407,34 @@ Shader::ShaderObjects::ShaderObjects(const osg::Shader* shader, unsigned int con
 
 Shader::PerContextShader* Shader::ShaderObjects::getPCS(const std::string& defineStr) const
 {
-    for(PerContextShaders::const_iterator itr = _perContextShaders.begin();
```
Missing GLExtensions::Get( false issue again.

tgeorgeVTMak commented on src/osg/Shader.cpp:
```diff
@@ -753,8 +765,9 @@ void Shader::PerContextShader::compileShader(osg::State& state)
-        _extensions->debugObjectLabel(GL_SHADER, _glShaderHandle, _shader->getName());
+        if (_extensions->glObjectLabel){
+           _extensions->glObjectLabel(GL_SHADER, _glShaderHandle, _shader->getName().length(), _shader->getName().c_str());
+        }
```
Missing change to _parseShaderDefines() probably another new change.

tgeorgeVTMak commented on src/osg/State.cpp:
```diff
@@ -158,6 +222,19 @@ State::~State()
+        //GLExtensions::Set(_contextID, 0);
+        // VRV_PATCH
+
+        // MAK REVIEW: The above was commented out in makOSG 3.4.  The below is new of 3.6.  Are they compatible?
```
I can't think of any reason we should be deleting the GLExtensions (at least it is unclear that we can reasonably know when we should delete it) so lets get rid of the Get() and Set() code below.

tgeorgeVTMak commented on src/osg/State.cpp:
```diff
@@ -1114,7 +1232,7 @@ void State::disableAllVertexArrays()
 void State::dirtyAllVertexArrays()
 {
-    OSG_INFO<<"State::dirtyAllVertexArrays()"<<std::endl;
+    OSG_INFO << "State::dirtyAllVertexArrays()" << std::endl;
```
missing change in initializeExtensionProcs
```
    #ifndef _DEBUG
    setTimestampBits(0);
#endif
```

tgeorgeVTMak replied:
Pretty sure we want that still

tgeorgeVTMak commented on src/osg/Texture.cpp:
```diff
@@ -228,7 +234,7 @@ GLenum assumeSizedInternalFormat(GLint internalFormat, GLenum type)
 }
 
-bool isCompressedInternalFormatSupportedByTexStorage(GLint internalFormat)
+bool isCompressedInternalFormatSupportedByTexStorrage(GLint internalFormat)
```
We should fix this typo instead of putting it back into makOsg 3.6

tgeorgeVTMak commented on src/osg/Texture.cpp:
```diff
@@ -2083,49 +2380,137 @@ bool Texture::areAllTextureObjectsLoaded() const
     return true;
 }
 
-GLenum Texture::selectSizedInternalFormat(const osg::Image* image) const
```
Why was this removed?

tgeorgeVTMak replied:
OK, I think I have it all worked through. I think we want this new function and a few other changes to applyTexImage2D_load. I will see if I can add them all as suggestions in the comments

tgeorgeVTMak commented on src/osg/Texture.cpp:
```diff
@@ -2083,49 +2380,137 @@ bool Texture::areAllTextureObjectsLoaded() const
-            }
-            else
-            {
-                return assumeSizedInternalFormat((GLenum)image->getInternalTextureFormat(), (GLenum)image->getDataType());
```
bring back this whole function and add this here (just a comment)
```
// VRV_PATCH - Just a comment
//Watch out, this function just returns the first format 
//match which might not be correct (chooses GL_SRGB8 before 
//GL_RGB8 which messes up default Terrapage texture format of GL_RGB8)
```

tgeorgeVTMak commented on src/osg/Texture.cpp:
```diff
@@ -2310,69 +2720,130 @@ void Texture::applyTexImage2D_load(State& state, GLenum target, const Image* ima
             int width  = inwidth;
             int height = inheight;
 
-            GLenum texStoragesizedInternalFormat = extensions->isTextureStorageEnabled && extensions->isTexStorage2DSupported() && (_borderWidth==0) ? selectSizedInternalFormat(image) : 0;
```
New version of lines 2722-2847 (should have both the new changes from OSG 3.6 and the makOsg 3.4 changes. 
```
            GLenum texStoragesizedInternalFormat = extensions->isTextureStorageEnabled && extensions->isTexStorage2DSupported() && (_borderWidth==0) ? selectSizedInternalFormat(image) : 0;

            if (texStoragesizedInternalFormat!=0)
            {
                if (getTextureTarget()==GL_TEXTURE_CUBE_MAP)
                {
                    if (target==GL_TEXTURE_CUBE_MAP_POSITIVE_X)
                    {
                        // only allocate on first face image
                        extensions->glTexStorage2D(GL_TEXTURE_CUBE_MAP, numMipmapLevels, texStoragesizedInternalFormat, width, height);
                    }
                }
                else
                {
                    OsgProfileC("glTexStorage2D(alloc)", tracy::Color::Purple);
                    osg::CVSpan allocSpan(series_rt2, 0, "glTexStorage2D(alloc)");            
                    extensions->glTexStorage2D(target, numMipmapLevels, texStoragesizedInternalFormat, width, height);
                }

                if( !compressed_image )
                {
                   //VRV_PATCH
                   if (target != GL_TEXTURE_2D || !getTextureObjectManager(contextID)->getTextureStreamingActive())
                   {
                      OsgProfileC("sendMipmaps", tracy::Color::Purple);
                      osg::CVSpan UpdateTick(series_rt2, 4, "sendMipmaps");
                      for( GLsizei k = 0 ; k < numMipmapLevels  && (width || height) ;k++)
                      {

                        if (width == 0)
                            width = 1;
                        if (height == 0)
                            height = 1;

                        glTexSubImage2D( target, k,
                            0, 0,
                            width, height,
                            (GLenum)image->getPixelFormat(),
                            (GLenum)image->getDataType(),
                            dataPtr + image->getMipmapOffset(k));

                        width >>= 1;
                        height >>= 1;
                      }
                   }
                   //VRV_PATCH
                   else {
                      // start streaming them
                      textureObject->_currentMipMapToApply = numMipmapLevels-1;
                   }
                }
                else if (extensions->isCompressedTexImage2DSupported())
                {
                   //VRV_PATCH
                   if (target != GL_TEXTURE_2D || !getTextureObjectManager(contextID)->getTextureStreamingActive())
                   {
                      OsgProfileC("sendCompressedMipmaps", tracy::Color::Purple);
                      osg::CVSpan UpdateTick(series_rt2, 4, "sendCompressedMipmaps");

                      GLint blockSize,size;
                      for( GLsizei k = 0 ; k < numMipmapLevels  && (width || height) ;k++)
                      {
                        if (width == 0)
                            width = 1;
                        if (height == 0)
                            height = 1;

                        getCompressedSize(image->getInternalTextureFormat(), width, height, 1, blockSize,size);

                        //state.checkGLErrors("before extensions->glCompressedTexSubImage2D(");

                        extensions->glCompressedTexSubImage2D(target, k,
                            0, 0,
                            width, height,
                            (GLenum)image->getPixelFormat(),
                            size,
                            dataPtr + image->getMipmapOffset(k));

                        //state.checkGLErrors("after extensions->glCompressedTexSubImage2D(");

                        width >>= 1;
                        height >>= 1;
                      }
                   }
                   else {
                      // start streaming them
                      textureObject->_currentMipMapToApply = numMipmapLevels-1;
                   }
                }

```

tgeorgeVTMak commented on src/osg/Texture.cpp:
```diff
@@ -2749,10 +3225,12 @@ Texture::GenerateMipmapMode Texture::mipmapBeforeTexImage(const State& state, bo
+            }
         }
 
-        glTexParameteri(getTextureTarget(), GL_GENERATE_MIPMAP_SGIS, GL_TRUE);
```
I don't think we want this change. I think the 3.6 version is good.

tgeorgeVTMak commented on src/osg/Texture.cpp:
```diff
@@ -2775,7 +3259,7 @@ void Texture::mipmapAfterTexImage(State& state, GenerateMipmapMode beforeResult)
         }
         case GENERATE_MIPMAP_TEX_PARAMETER:
-            glTexParameteri(getTextureTarget(), GL_GENERATE_MIPMAP_SGIS, GL_FALSE);
+            glTexParameteri(GL_TEXTURE_2D, GL_GENERATE_MIPMAP_SGIS, GL_FALSE);
```
Same with this one. I think we want the OSG 3.6 version.

tgeorgeVTMak commented on src/osg/Texture1D.cpp:
```diff
@@ -211,7 +211,7 @@ void Texture1D::apply(State& state) const
 
         applyTexParameters(GL_TEXTURE_1D,state);
 
-        // update the modified count to show that it is up to date.
```
Where did these changes come from? I am not seeing them in makOsg 3.4

tgeorgeVTMak replied:
This comment refers to the whole file.

tgeorgeVTMak replied:
Nevermind, I see that it is because selectSizedInternalFormat was removed from texture.cpp. I think now that I have a new version of texture.cpp specified in the comments for it then we can put these changes back in.

tgeorgeVTMak commented on src/osg/Texture2D.cpp:
```diff
@@ -243,7 +419,6 @@ void Texture2D::apply(State& state) const
     }
     else if (_image.valid() && _image->data())
     {
-        GLExtensions * extensions = state.get<GLExtensions>();
```
More code that can go back in now that selectSizedInternalFormat is going back into Texture.cpp

tgeorgeVTMak commented on src/osg/Texture2D.cpp:
```diff
@@ -297,30 +471,18 @@ void Texture2D::apply(State& state) const
+        applyTexParameters(GL_TEXTURE_2D,state);
 
         // no image present, but dimensions at set so lets create the texture
-         GLExtensions * extensions = state.get<GLExtensions>();
```
More code that can go back in now that selectSizedInternalFormat is going back into Texture.cpp

tgeorgeVTMak commented on src/osg/Texture2DArray.cpp:
```diff
@@ -336,53 +336,42 @@ void Texture2DArray::apply(State& state) const
         // compute the internal texture format, this set the _internalFormat to an appropriate value.
         computeInternalFormat();
 
-        GLenum texStorageSizedInternalFormat = extensions->isTextureStorageEnabled && (_borderWidth==0) ? selectSizedInternalFormat(_images[0].get()) : 0;
```
I think this whole file can go back to the 3.6 version now that selectSizedInternalFormat is going back into Texture.cpp

tgeorgeVTMak commented on src/osg/Texture2DMultisample.cpp:
```diff
@@ -102,32 +102,29 @@ void Texture2DMultisample::apply(State& state) const
     }
     else if ( (_textureWidth!=0) && (_textureHeight!=0) && (_numSamples!=0) )
     {
-        // no image present, but dimensions at set so lets create the texture
```
I think this whole file can go back to the 3.6 version now that selectSizedInternalFormat is going back into Texture.cpp

tgeorgeVTMak commented on src/osg/Texture2DMultisampleArray.cpp:
```diff
@@ -0,0 +1,150 @@
+/* -*-c++-*- OpenSceneGraph - Copyright (C) 1998-2010 Robert Osfield
```
We may want to incorporate the new code for 3.6 that is in Texture2DMultisample.cpp into this file as well.

tgeorgeVTMak commented on src/osg/Texture3D.cpp:
```diff
@@ -315,30 +315,20 @@ void Texture3D::apply(State& state) const
     }
     else if ( (_textureWidth!=0) && (_textureHeight!=0) && (_textureDepth!=0) && (_internalFormat!=0) )
     {
-        // no image present, but dimensions at set so lets create the texture
```
think this whole file can go back to the 3.6 version now that selectSizedInternalFormat is going back into Texture.cpp

tgeorgeVTMak commented on src/osg/TextureCubeMap.cpp:
```diff
@@ -293,12 +293,8 @@ void TextureCubeMap::apply(State& state) const
             _textureWidth = _textureHeight = minimum( _textureWidth , _textureHeight );
         }
 
-        GLenum texStorageSizedInternalFormat = extensions->isTextureStorageEnabled && (_borderWidth==0) ? selectSizedInternalFormat(_images[0].get()) : 0;
```
think this whole file can go back to the 3.6 version now that selectSizedInternalFormat is going back into Texture.cpp

tgeorgeVTMak commented on src/osg/TextureRectangle.cpp:
```diff
@@ -231,24 +231,19 @@ void TextureRectangle::apply(State& state) const
     }
     else if (_image.valid() && _image->data())
     {
-        GLExtensions * extensions = state.get<GLExtensions>();
```
think this whole file can go back to the 3.6 version now that selectSizedInternalFormat is going back into Texture.cpp

tgeorgeVTMak commented on src/osgDB/Registry.cpp:
```diff
@@ -1171,7 +1168,7 @@ ReaderWriter::ReadResult Registry::read(const ReadFunctor& readFunctor)
             options->setDatabasePath(archiveName);
 
-            osg::ref_ptr<ReadFunctor> rf(readFunctor.cloneType(fileName, options.get()));
+            std::auto_ptr<ReadFunctor> rf(readFunctor.cloneType(fileName, options.get()));
```
This is a strange change. Why was it done? I am not seeing the same change in makOsg 3.4

tgeorgeVTMak replied:
Or at least it seems like we should take the 3.6 change in this case instead of using the old 3.4 way.

tgeorgeVTMak commented on src/osgParticle/Particle.cpp:
```diff
@@ -101,7 +101,9 @@ bool osgParticle::Particle::update(double dt, bool onlyTimeStamp)
 
     //Compute the current texture tile based on our normalized age
-    int currentTile = _start_tile + static_cast<int>(x * getNumTiles());
+    // BEGIN VRV_PATCH
```
It is unclear if all the particle changes made it in or if they are even needed anymore. I think we are good if the old OSG particle systems look ok in VRV and not worry about the performance improvements since we have our own particle system that is separate from OSG now.

tgeorgeVTMak commented on src/osgPlugins/OpenFlight/AncillaryRecords.cpp:
```diff
@@ -276,23 +355,20 @@ class UVList : public Record
             uint32 mask = in.readUInt32(0);
 
             int numLayers = bitCount(mask);
-            if (numLayers>0)
```
I think we want these changes. It is just checking to make sure the number of layers is greater than 0.

tgeorgeVTMak commented on src/osgPlugins/OpenFlight/AttrData.cpp:
```diff
@@ -92,90 +92,13 @@ AttrData::AttrData() :
     hemisphere(1),
     comments(""),
     attrVersion(0),
-    controlPoints(0),
```
I am not seeing these changes in the 3.4 branch. Are we good to just use the 3.6 version?

tgeorgeVTMak commented on src/osgPlugins/OpenFlight/FltExportVisitor.cpp:
```diff
@@ -425,9 +425,7 @@ FltExportVisitor::apply( osg::Geode& node )
                 else if ( (prim->getType() == osg::PrimitiveSet::DrawElementsUBytePrimitiveType) ||
                         (prim->getType() == osg::PrimitiveSet::DrawElementsUShortPrimitiveType) ||
                         (prim->getType() == osg::PrimitiveSet::DrawElementsUIntPrimitiveType) )
-                {
```
lets put these back in.

tgeorgeVTMak commented on src/osgPlugins/OpenFlight/FltExportVisitor.h:
```diff
@@ -184,10 +184,10 @@ class FltExportVisitor : public osg::NodeVisitor
-    osg::ref_ptr<TexturePaletteManager>      _texturePalette;
-    osg::ref_ptr<LightSourcePaletteManager>  _lightSourcePalette;
-    osg::ref_ptr<VertexPaletteManager>       _vertexPalette;
+    std::auto_ptr<MaterialPaletteManager>     _materialPalette;
```
Lets use the 3.6 version of this file as well. We don't have any 3.4 changes to this file.

tgeorgeVTMak commented on src/osgPlugins/OpenFlight/GeometryRecords.cpp:
```diff
@@ -35,1434 +36,1495 @@
-    }
-}
-
-void addDrawableAndReverseWindingOrder( osg::Geode* geode )
```
Let's take the 3.6 version of this function since it is the same change we made just with different variable names.

john-terrell replied:
Reverted change to this function (pending)

tgeorgeVTMak commented on src/osgPlugins/OpenFlight/LightSourcePaletteManager.h:
```diff
@@ -33,7 +33,7 @@ namespace flt
 
 
-class LightSourcePaletteManager : public osg::Referenced
+class LightSourcePaletteManager
```
Let's take this osg 3.6 change

john-terrell replied:
Reverted change to this file (pending)

tgeorgeVTMak commented on src/osgPlugins/OpenFlight/MaterialPaletteManager.h:
```diff
@@ -32,7 +32,7 @@ namespace flt
 class DataOutputStream;
 
 
-class MaterialPaletteManager : public osg::Referenced
```
Let's take the 3.6 version of this file.

john-terrell replied:
Reverted change to this file (pending)

tgeorgeVTMak commented on src/osgPlugins/OpenFlight/Pools.cpp:
```diff
@@ -125,7 +125,7 @@ osg::Material* MaterialPool::getOrCreateMaterial(int index, const osg::Vec4& fac
 
     // Create final material.
-    osg::Material* material = osg::clone(poolMaterial, osg::CopyOp::SHALLOW_COPY);
+    osg::Material* material = dynamic_cast<osg::Material*>(poolMaterial->clone(osg::CopyOp::SHALLOW_COPY));
```
I think we can use the 3.6 version of this file.

john-terrell replied:
Reverted changes to this file (pending)

tgeorgeVTMak commented on src/osgPlugins/OpenFlight/ReaderWriterATTR.cpp:
```diff
@@ -185,7 +185,7 @@ ReaderWriter::ReadResult ReaderWriterATTR::readObject(const std::string& file, c
 
 ReaderWriter::WriteResult
-ReaderWriterATTR::writeObject(const osg::Object& object, const std::string& fileName, const Options* /*options*/) const
+ReaderWriterATTR::writeObject(const osg::Object& object, const std::string& fileName, const Options* options) const
```
we can probably use the 3.6 version of this file.

john-terrell replied:
Reverted changes to this file (pending)

tgeorgeVTMak commented on src/osgPlugins/OpenFlight/ReaderWriterFLT.cpp:
```diff
@@ -1,40 +1,44 @@
+#include "DtCDBSigSizeTable.h"
 
-#define SERIALIZER() OpenThreads::ScopedLock<OpenThreads::ReentrantMutex> lock(_serializerMutex)
+#define SERIALIZER() OpenThreads::ScopedLock<OpenThreads::ReentrantMutex> lock(_serializerMutex)  
```
We should take the 3.6 version of this change if we can.

john-terrell replied:
WIll take this change excepting anything in the following comments for this file.

john-terrell replied:
Pending.

tgeorgeVTMak commented on src/osgPlugins/OpenFlight/ReaderWriterFLT.cpp:
```diff
@@ -174,7 +178,7 @@ class ReadExternalsVisitor : public osg::NodeVisitor
 
             // read external
-            osg::ref_ptr<osg::Node> external = osgDB::readRefNodeFile(filename,_options.get());
+            osg::ref_ptr<osg::Node> external = osgDB::readNodeFile(filename,_options.get());
```
We should take the 3.6 version of this change if we can. This is one that I am more hesitant on though since readNodeFile might go through a different path in the VRV file loader vs readRefNodeFile

john-terrell replied:
I made a comment in the source that this was changed for OSG 3.6.   Folks debugging this will see there's a change.

tgeorgeVTMak commented on src/osgPlugins/OpenFlight/ReaderWriterFLT.cpp:
```diff
@@ -373,15 +411,18 @@ class FLTReaderWriter : public ReaderWriter
                 document.setUseTextureAlphaForTransparancyBinning(options->getOptionString().find("noTextureAlphaForTransparancyBinning")==std::string::npos);
                 OSG_DEBUG << readerMsg << "noTextureAlphaForTransparancyBinning=" << !document.getUseTextureAlphaForTransparancyBinning() << std::endl;
 
-                document.setReadObjectRecordData(options->getOptionString().find("readObjectRecordData")!=std::string::npos);
```
I think we want this fix since it appears to be a typo fix in the 3.6 code.

john-terrell replied:
Pending.

tgeorgeVTMak commented on src/osgPlugins/OpenFlight/ReaderWriterFLT.cpp:
```diff
@@ -593,7 +697,7 @@ class FLTReaderWriter : public ReaderWriter
             }
 
             flt::DataOutputStream dos( fOut.rdbuf(), fltOpt->getValidateOnly() );
-            flt::FltExportVisitor fnv( &dos, fltOpt.get() );
```
I think we want this change from 3.6

john-terrell replied:
Pending.

tgeorgeVTMak commented on src/osgPlugins/OpenFlight/Record.cpp:
```diff
@@ -84,6 +84,14 @@ void flt::insertMatrixTransform(osg::Node& node, const osg::Matrix& matrix, int
     osg::ref_ptr<osg::Node> ref = &node;
     osg::Node::ParentList parents = node.getParents();
 
+    // Disconnect node from parents.
```
I am not seeing any changes to this file in makOsg 3.4 so we probably want the 3.6 version of this file.

john-terrell replied:
Replaced file with OSG 3.6 version.

tgeorgeVTMak commented on src/osgPlugins/OpenFlight/Registry.h:
```diff
@@ -1,13 +1,13 @@
-/* -*-c++-*- OpenSceneGraph - Copyright (C) 1998-2006 Robert Osfield
```
This file appears to be keeping around a bunch of stuff that was removed in 3.6, are we sure we need all of the stuff we kept around?

tgeorgeVTMak commented on src/osgPlugins/OpenFlight/TexturePaletteManager.h:
```diff
@@ -35,7 +35,7 @@ class DataOutputStream;
 class FltExportVisitor;
 
 
-class TexturePaletteManager : public osg::Referenced
```
Let's use the 3.6 version

john-terrell replied:
Pending.

tgeorgeVTMak commented on src/osgPlugins/OpenFlight/VertexPaletteManager.cpp:
```diff
@@ -79,8 +79,8 @@ VertexPaletteManager::add( const osg::Geometry& geom )
 
-    const bool cpv =( osg::getBinding(geom.getColorArray()) == osg::Array::BIND_PER_VERTEX );
-    const bool npv =( osg::getBinding(geom.getNormalArray()) == osg::Array::BIND_PER_VERTEX );
+    const bool cpv =( geom.getColorBinding() == osg::Geometry::BIND_PER_VERTEX );
```
These two lines are the only lines changed in makOsg 3.4, maybe they should be the only 2 lines we keep changed in 3.6

tgeorgeVTMak replied:
Wait, the const for write() is another change.

john-terrell replied:
Will keep those two lines and replace the rest with 3.6 versions.   Will keep write() as const unless something in there is trying to modify member data.   Will address on next build.

john-terrell replied:
Will use the above two lines and revert the rest.   The const on write will be removed as well since verticesStr will be non-mutable (per next comment).

tgeorgeVTMak commented on src/osgPlugins/OpenFlight/VertexPaletteManager.h:
```diff
@@ -53,7 +54,7 @@ class VertexPaletteManager : public osg::Referenced
 
     unsigned int byteOffset( unsigned int idx ) const;
 
-    void write( DataOutputStream& dos ) const;
```
This and the removal of the mutable on verticesStr are the only change we should take. The rest of the changes we should take the 3.6 version instead.

john-terrell replied:
Roger.  Against previous comment in the .cpp file, I'll remove mutable on verticesStr as well as const on write().   Everything else will be 3.6 changes.

john-terrell replied:
Will revert changes and:
1) Remove const from write().
2) Remove mutable from verticesStr().

tgeorgeVTMak commented on src/osgPlugins/OpenFlight/expGeometryRecords.cpp:
```diff
@@ -50,7 +50,7 @@ static unsigned int LAYER_7( 0x80000000 >> 6 );
 
 
 bool
-FltExportVisitor::isLit( const osg::Geometry& /*geom*/ ) const
```
I think we want the 3.6 version of this file. There are no changes in 3.4 makOsg

john-terrell replied:
Reverted changes to this file (pending)

tgeorgeVTMak commented on src/osgPlugins/OpenFlight/expPrimaryRecords.cpp:
```diff
@@ -688,13 +688,13 @@ FltExportVisitor::writeLightPoint( const osgSim::LightPointNode* lpn )
     float32 horizLobe( 360.f );
     float32 vertLobe( 360.f );
     float32 lobeRoll( 0.f );
-    const osgSim::DirectionalSector* dirSector = dynamic_cast< osgSim::DirectionalSector* >( lp0._sector.get() );
```
Looks like the 3.6 version fixed the same problem we did but in a better way. Let's use the 3.6 version.

john-terrell replied:
Reverted.

tgeorgeVTMak commented on src/osgPlugins/fbx/CMakeLists.txt:
```diff
@@ -18,27 +21,22 @@ SET(TARGET_H
+    fbxMatrixTransformVisitor.h
 )
 
-IF(CMAKE_COMPILER_IS_GNUCXX)
```
It is unclear that we are doing the right thing here. I think we may want both the 3.6 changes and our changes here. Maybe it should look like this:
```
INCLUDE_DIRECTORIES(${BOOST_INCLUDE_DIR})

link_directories(${BOOST_LIB_DIR})

IF(CMAKE_COMPILER_IS_GNUCXX)
SET(CMAKE_CXX_FLAGS "-w")
ELSEIF(CMAKE_CXX_COMPILER_ID STREQUAL "Clang")
    REMOVE_CXX_FLAG(-pedantic)
    REMOVE_CXX_FLAG(-Wall)
    REMOVE_CXX_FLAG(-Wunused)
    REMOVE_CXX_FLAG(-Wextra)
ENDIF()

IF(WIN32)
    OPTION(FBX_SHARED OFF)
    IF(FBX_SHARED)
        ADD_DEFINITIONS(-DFBXSDK_SHARED)
    ENDIF(FBX_SHARED)

    SET(TARGET_EXTERNAL_LIBRARIES wininet)
ENDIF(WIN32)

```

tgeorgeVTMak commented on src/osgPlugins/fbx/WriterCompareTriangle.cpp:
```diff
@@ -1,9 +1,10 @@
+                                             :    geode(geode)
 {
-    cutscene(in_nbVertices, geode.getDrawable(0)->asGeometry()->getBoundingBox());
+    cutscene(nbVertices, geode.getDrawable(0)->asGeometry()->getBoundingBox());
```
Looks like we should use the 3.6 version of this file.

john-terrell replied:
Reverted changes to this file (pending)

tgeorgeVTMak commented on src/osgPlugins/fbx/WriterNodeVisitor.cpp:
```diff
@@ -444,7 +444,7 @@ WriterNodeVisitor::setLayerTextureAndMaterial(FbxMesh* mesh)
 }
 
 void
-WriterNodeVisitor::setControlPointAndNormalsAndUV(const GeometryList& geometryList,
```
I am not seeing any change to this file in makOsg 3.4 so we should probably use the 3.6 version of it.

john-terrell replied:
Resolved.

tgeorgeVTMak commented on src/osgPlugins/fbx/WriterNodeVisitor.h:
```diff
@@ -45,8 +45,8 @@ struct Triangle
 
 struct VertexIndex
 {
```
I am not seeing any changes to this file in 3.4 so we should probably take the 3.6 version of the file.

john-terrell replied:
Reverted changes to this file (pending)

tgeorgeVTMak commented on src/osgPlugins/ive/ReadWrite.h:
```diff
@@ -179,5 +192,12 @@ namespace ive {
 #define IVETEXT3D                       0x10000002
 #define IVEFADETEXT                     0x10000003
 
+class ReadWrite{
```
We probably don't need this class added since we didn't modify it in the 3.4 makOsg.

john-terrell replied:
Removed the extraneous class.

john-terrell replied:
Update: this class is still in use:

src\osgplugins\ive\ExtendedMaterial.h(8): error C2504: 'ReadWrite': base class undefined
src\osgplugins\ive\AnimationMatrixTransform.h(12): error C2504: 'ReadWrite': base class undefined
src\osgplugins\ive\Animation.h(13): error C2504: 'ReadWrite': base class undefined

tgeorgeVTMak commented on src/osgPlugins/nvtt/NVTTImageProcessor.cpp:
```diff
@@ -4,7 +4,9 @@
 
 #include <nvtt/nvtt.h>
 #include <string.h>
-
```
We should double check that NVTT project is building without the 3.4 makOsg changes to the cmakelists.txt file.

tgeorgeVTMak commented on src/osgPlugins/txp/trpage_io.h:
```diff
@@ -299,7 +299,7 @@ TX_EXDECL class TX_CLDECL trpg3dPoint {
 TX_EXDECL class TX_CLDECL trpgColor {
 public:
     trpgColor(float64 r,float64 g,float64 b) {red = r; green = g; blue = b;}
-    trpgColor(void) : red(1.0), green(1.0), blue(1.0) { };
```
Let's use the 3.6 default for this instead of the vrvantage specified default.

john-terrell replied:
Resolved.

tgeorgeVTMak commented on src/osgPlugins/txp/trpage_managers.cpp:
```diff
@@ -756,7 +758,6 @@ trpgPageManager::trpgPageManager()
     lastLoad = None;
     lastLod = 0;
     lastTile = 0;
-    scale = 1.0;
```
Lets take the 3.6 version of this line

john-terrell replied:
Readded that line.   Rest of file as-is.

tgeorgeVTMak commented on src/osgPlugins/txp/trpage_readbuf.cpp:
```diff
@@ -721,5 +721,4 @@ trpgrAppFileCache::OpenFile::OpenFile()
     id = -1;
     row = 0;
     col = 0;
-
```
lets just take the 3.6 version of this file.

john-terrell replied:
Reverted changes to this file (pending)

tgeorgeVTMak commented on src/osgText/Text.cpp:
```diff
@@ -1064,12 +1064,6 @@ void Text::computeColorGradientsPerCharacter()
         {
             case 0: // top-left
                 {
-                    (*colorCoords)[i] = _colorGradientTopLeft;
```
Why was this removed?

john-terrell replied:
Good catch.   Bad auto-merge.   Fixed.

tgeorgeVTMak commented on src/osgUtil/MeshOptimizers.cpp:
```diff
@@ -498,7 +498,9 @@ float findVertexScore (Vertex& vert)
         return -1.0f;
     }
     float score = 0.0f;
-    int cachePosition = vert.cachePosition;
```
It looks like LineSegmentIntersector.cpp is missing from the diff. It does appear to have changed drastically between 3.4 and 3.6 and there are other significant changes that we made in 3.4 makOsg. I am not sure which version or combinations of the versions we should use. Probably worth a defect in ontime to track the difference and continue to use the 3.6 version for now.

john-terrell replied:
This is one case where the merge wasn't feasible.   The changes looked significant enough that I opted for the 3.6 versions.   I can open an OnTime issue in case issues pop up.

john-terrell replied:
Tracked as OnTime issue 5839 (subitem of 5327)

tgeorgeVTMak commented on src/osgUtil/Optimizer.cpp:
```diff
@@ -34,6 +34,8 @@
 #include <osg/Timer>
 #include <osg/TexMat>
 #include <osg/io_utils>
+//VRV_PATCH#
```
We need to double check that any changes to this file are also made to DtOptimizer in VRV.

tgeorgeVTMak commented on src/osgViewer/ViewerBase.cpp:
```diff
@@ -205,11 +205,13 @@ void ViewerBase::setThreadingModel(ThreadingModel threadingModel)
 {
     if (_threadingModel == threadingModel) return;
 
+    bool needSetUpThreading = _threadsRunning;
```
where did the changes in setThreadingModel come from?

john-terrell replied:
I don't see a change in either 3.6 or makOSG 3.4.   Bad auto-merge?   Reverted to the OSG 3.6 version.

tgeorgeVTMak commented on src/osgWrappers/serializers/osg/HeightField.cpp:
```diff
@@ -31,18 +31,22 @@ static bool checkHeights( const osg::HeightField& shape )
 
 static bool readHeights( osgDB::InputStream& is, osg::HeightField& shape )
 {
-    osg::ref_ptr<osg::Array> array = is.readArray();
```
lets change the FloatArray to be called farray to match OSG 3.6

john-terrell replied:
Fixed

john-terrell committed:
TimG PR feedback #1

john-terrell committed:
Fixed broken auto-merge in Text.cpp

john-terrell committed:
Fix for bad merge.

john-terrell committed:
TimG feedback.

john-terrell committed:
Missed variable name change.

john-terrell committed:
Fixes for feedback changes.

john-terrell committed:
Reverted feedback change.  Still used.

john-terrell committed:
pzimmons (Rev 208969) - Now that we have done our TA2 IPB4, etc. branches. Updating ActiveUniformMap, UniformMap to a flat_map (vector based map) instead of using std::map.

A flat_map is faster than a std::map for a small number of elements and slower than a std::unordered_map for a large number of items
A flat_map is sorted which is required for osg (there is also an unordered_flat_map which is unordered).

john-terrell committed:
tgeorge (Rev 209269)
Feature 5831 - Split the linkProgram up into 3 functions so it can avoid compiling the shaders if a program binary is loaded correctly.
https://ontime/viewitem.aspx?id=5831&type=features&force_use_number=false

john-terrell committed:
Fixed missing <cr>

john-terrell committed:
Fixed bad merge.

john-terrell committed:
rfranco (Rev 209456) - Initial GraphicsContext

john-terrell committed:
rfranco (Rev 209545) - Defect 67686 Fix - Add a way to enable/disable a camera in a osg::GraphicsContext - which means adding/removing it from the class's internal list. When removing, we do not clean up anything associated with the camera which makes it different than GraphicsContext::removeCamera(...).

john-terrell committed:
Reverted OSGText to OSG 3.6 baseline

john-terrell committed:
Missed source for previous fix

john-terrell committed:
John/textfixes (#3)

* Reverted changes to Drawable.cpp

* Test for broken change.

* reverted Texture2D.cpp

* Added missing function from OSG 3.6

* More fixes for reverting Texture2D.

* Enabled OSG_GL_FIXED_FUNCTION_AVAILABLE

* WIP

* WIP

* Readded building of include/osg/GL instead of hardcoded.

* Removed unnecessary change in DisplaySettings.cpp

* WIP

* Updated Image.cpp

* Readded commenting out of conditional for state.convertShaderSourcetoOsgBuiltins()

* Revert of previous OSG 3.6 change.   Breaks textures.

* Removed core profile deprecated items.

* More Core Profile cleanup

* Texture fixes

* Readded VRV specific patch to Material::apply()

Co-authored-by: John W. Terrell <john.terrell@gmail.com>

john-terrell committed:
plavoie (Rev: 210092) - Defect #67974 - When a FBX model contain a key frame animation somehow its messing up the intersection in vantage/VRF. Since we don't use Key frame animation for now in vantage, I am disabling it for now until it will be added to indirect rendering for instancing

john-terrell committed:
tcarellas (Rev 210693) - OTF5895 - Merged vnc plugin branch into TOT makOSG
The STE plugins need that plugin for connecting to a VNC server in VR-Engage.
https://ontime/viewitem.aspx?id=5895&type=features&force_use_number=false
----
Modified : /makOSG/trunk
Modified : /makOSG/trunk/CMakeLists.txt
Modified : /makOSG/trunk/local.cmake
Modified : /makOSG/trunk/patches/3.4.0/src/osgPlugins/CMakeLists.txt

john-terrell committed:
tgeorge (Rev: 211330) - Defect 68456 - Added an immutable flag to the texture objects. This starts out false and can never be set back to false if it is ever set to true. VRV uses this to let OSG know that a texture has been made resident and should not be reused.
https://ontime/viewitem.aspx?id=68456&type=defects&force_use_number=false

john-terrell committed:
tcarellas (Rev: 211480) - Merged revision(s)  from makOSG/working/tcarellas-boost163:

john-terrell committed:
tgeorge (Rev: 212255) - Defect 68456 - Fixed an issue with an iterator getting used after it is invalid. Fixes a crash on linux.

john-terrell committed:
plavoie (Rev: 212390) - Feature 6038 - Support Diffuse Multi-texture (add and modulate)

john-terrell committed:
plavoie (Rev: 212431) - Feature 6038 - Support Diffuse Multi-texture adding (replace and decal)

john-terrell committed:
Fixed build break with src/osg/Texture.cpp

john-terrell committed:
WIP

john-terrell committed:
tgeorge (Rev 212557) - VRV-1574 - Added Reference tracking to osg::Reference. There is now a set of osg::Reference* to keep track of what objects are currently created. This will allow for tracking down orphaned pointers and memory leaks. Disabled by default uncomment line 146 to enable it.

john-terrell committed:
tgeorge (Rev 212680) - VRV-1565 - Moved Referenced.cpp to the top of the list of cpp files in order to have the global static mutex be the first thing created and the last thing destroyed. Fixes some memory access issues on exit.

john-terrell committed:
plavoie (Rev 212792) - Fix for crash introduced by the multi-texture on certain models with no multi-texture

john-terrell committed:
plavoie (Rev 212879) - Fix problem introduced by multitexture on non-multitexture models (having a background color)

john-terrell committed:
Added plugin logging support

john-terrell committed:
Small output fix for previous commit.

john-terrell committed:
Added a few missing cases for plugin logging.

john-terrell committed:
WIP: Fixes.

john-terrell committed:
Fixed to not send full filenames in certain circumstances to the plugin log.

john-terrell committed:
WIP:More logging work

john-terrell committed:
Modified FindCOLLADA/FindFreeType to look for Mak specific versions.   Updated OpenFlight and fbx plugins to find boost packages since they require them.

john-terrell committed:
Added support for Tracy profilng

john-terrell committed:
Updated FindTracy.cmake to only look for libraries on Windows

john-terrell committed:
tgeorge (Rev 216277) - VRV-1851 - Fixed Texture2DArrays to correctly send the images down to the GPU after a graphics context recreation.

john-terrell committed:
pprabhu (Rev 216568) - patch to notify on texture parameter re-application/dirtying when texture objects are resident

john-terrell committed:
Fix to ensure VertexArrayState is sized correctly relative to the geometry's _vertexAttribList.   This fixes a crash when drawing DiGuys.

john-terrell committed:
Fix to ensure VertexArrayState is sized correctly relative to the geometry's _vertexAttribList.   This fixes a crash when drawing DiGuys.

john-terrell committed:
Merge branch 'mak/osg-3.4-merge' of github.com:VTMAK/OpenSceneGraph into mak/osg-3.4-merge

john-terrell committed:
Re-added mipmap sending.

john-terrell committed:
tgeorge (Rev 217174) - VRV-1909 - Fix makOsg to build with Tracy on linux.

john-terrell committed:
Added CMake find package for Concurrency Macros (CVMacros).   Updated FindTracy to use latest lib name.

john-terrell committed:
Added additional CMake find script for CDBAPI.   Modified OpenFlight to use it.   Fixed missing COLLADA include dir to its CMake script.

john-terrell committed:
Updated root CMakeLists to properly scope Trace package to Windows only.

Commit:
-fixes for F2 stats/osg::Text
Steve Koren:
"The basic problem was happening because glBindVertexArray was unbinding the GL_ELEMENT_ARRAY_BUFFER binding point, and OSG was unaware of that, so it didn't dirty its internal cache of what was bound, which in turn meant that the right thing wasn't getting bound later on because OSG didn't know glBindVertexArray altered the binding point."

Commit:
-fixes for di-guys
 -VertexArrayState
  -legitimate and correct fix for checking refence count
  -legitimate and correct fix for resizing a list correctly
 -State.cpp
  -fix geometry vbo, vao settings to match the same as when display settings have VBO hint (plus comments)
 -Geometry.cpp
  -set the correct vertex pointer data type when binding per vertex

Commit:
-fix heap corruption because of incorrect double deletion

Commit:
-fix for mipmaps from makOsg/trunk

Commit:
-osg changed the way errors are reported in 3.6. If a file is not loaded, it prints out the wrong error message as a warning
-This patch against 3.6+ fixes that. It reports exactly what went wrong if a file was not loaded at the appropriate level
 (error reading: warning, not enough memory: warning, not found: info)

Commit:
-clean up earlier patch for Geometry
-patch for fbx:
 Revision: 219328
Author: plavoie
Date: Wednesday, October 21, 2020 8:26:08 AM
Message:
VRV-2254 - Fix UV on multi diffuse texture
----
Modified : /makOSG/trunk/patches/3.4.0/src/osgPlugins/fbx/fbxRMesh.cpp

Commit:
-fixes for:
https://jira.mak.com:8443/browse/VRV-2282
Renable line stippling for now. Please also see:
https://jira.mak.com:8443/browse/VRV-2287

Commit:
-fixes for glPolygonStipple
-updated comments
